src/Agent.Listener/DistributedTask.Pipelines/Yaml/TypeConverters/ConverterUtil.variables.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using Microsoft.TeamFoundation.DistributedTask.Orchestration.Server.Pipelines.Yaml.Contracts;
+using YamlDotNet.Core;
+using YamlDotNet.Core.Events;
+using YamlDotNet.Serialization;
+
+namespace Microsoft.TeamFoundation.DistributedTask.Orchestration.Server.Pipelines.Yaml.TypeConverters
+{
+    internal static partial class ConverterUtil
+    {
+        internal static IList<IVariable> ReadVariables(IParser parser, Boolean simpleOnly = false)
+        {
+            var result = new List<IVariable>();
+            if (parser.Accept<MappingStart>())
+            {
+                // The simple syntax is:
+                //   variables:
+                //     var1: val1
+                //     var2: val2
+                foreach (KeyValuePair<String, String> pair in ReadMappingOfStringString(parser, StringComparer.OrdinalIgnoreCase))
+                {
+                    result.Add(new Variable() { Name = pair.Key, Value = pair.Value });
+                }
+            }
+            else
+            {
+                // When a variables template is referenced, sequence syntax is required:
```

Comment:
a list of objects is still required when using a template (due to syntax `- template: foo`)